File: WRORobotRCM/WRORobotRCM/Action.cpp
#include "Action.h"
#include "Robot.h"
Action::Action(vector<shared_ptr<Request>> requests)
{
	this->requests = requests;
	requestsWorking = vector<bool>();
	for (int i = 0; i < requests.size(); i++)
	{
		requestsWorking.push_back(true);
	}
	haveInit = false;
}
bool Action::Update(Robot& robot)
{
	bool working = false;
	for (int i = 0; i < requests.size(); i++)
	{
		shared_ptr<Request> request = requests[i];
		if (requestsWorking[i])
		{
			if (!haveInit)
			{
				robot.Components[request->Motor].StartRequest(request, robot);
			}
			requestsWorking[i] = robot.Components[request->Motor].Update(robot);
			if (requestsWorking[i])
			{
				working = true;
			}
		}
	}
	haveInit = true;
	return working;
}

File: WROSimulatorV2/WROSimulatorV2/Action.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WROSimulatorV2
{
    public class Action
    {
        List<Request> requests;
        List<bool> requestsWorking;
        bool haveInit;
        public Action(List<Request> requests)
        {
            requestsWorking = new List<bool>();
            this.requests = requests;
            HashSet<Motors> motors = new HashSet<Motors>();
            foreach(var r in requests)
            {
                requestsWorking.Add(true);
                if(r.Motor != Motors.NotApplicable)
                {
                    if(motors.Contains(r.Motor))
                    {
                        throw new IndexOutOfRangeException();
                    }
                    motors.Add(r.Motor);
                }
            }
            haveInit = false;
        }
        public bool Update(Robot robot, long elapsedMillis)//returns true if working false if done
        {
            bool working = false;
            for (int i = 0; i < requests.Count; i++)
            {
                Request request = requests[i];
                if (requestsWorking[i])
                {
                    if (!haveInit)
                    {
                        robot.Components[request.Motor].StartRequest(request, robot);
                    }
                    requestsWorking[i] = robot.Components[request.Motor].Update(robot,elapsedMillis);
                    if (requestsWorking[i])
                    {
                        working = true;
                    }
                }
            }
            haveInit = true;
            return working;
        }
    }

    
}
